python/pyarrow/array.pxi

```diff
@@ -1170,7 +1170,13 @@ cdef class Array(_PandasConvertible):
         array = PyObject_to_object(out)
 
         if isinstance(array, dict):
-            array = np.take(array['dictionary'], array['indices'])
+            if zero_copy_only or not self.null_count:
+                # zero_copy doesn't allow for nulls to be in the array
+                array = np.take(array['dictionary'], array['indices'])
+            else:
+                missings = array["indices"] < 0
+                array = np.take(array['dictionary'], array['indices'])
+                array[missings] = np.NaN
```

Comment:
This will not work for all data types, I think (eg a dictionary array with integer dictionary). Also, for strings, we might want to use `None` instead of `np.nan`, since this is what we do in basic conversion as well (eg see `pa.array(['a', None]).to_numpy(zero_copy_only=False)`). 

Given those complexities, I am wondering if it might not be easier to *first* convert to a "dense" array before doing the arrow->python conversion.

Reply:
Thanks for catching the `np.NaN` with ints, I actually noticed that yesterday but forgot to deal with it when I got back to this ticket.

Regarding the `None` vs `np.NaN`.
I think there is an inconsistency already, because I copied the behaviour from `to_pandas` and it seems that `to_pandas` already leads to two different results when used on `DictionaryArray`  or `Array`:

```
>>> pa.array(['a', None]).to_pandas().tolist()
['a', None]
>>> pa.array(['a', None]).to_numpy(zero_copy_only=False).tolist()
['a', None]
```

VS

```
>>> pa.DictionaryArray.from_arrays(pa.array([0, None]), pa.array(['a'])).to_pandas().tolist()
```

So it seems that for `DictionaryArray` we already used `NaN`, not `None`, which lead to the reason why I used `NaN`.

Should we uniform the behaviour and switch to `None` for `DictionaryArray.to_pandas` too as we do for `Array.to_pandas`?

Reply:
> So it seems that for `DictionaryArray` we already used `NaN`, not `None`, which lead to the reason why I used `NaN`.

That's actually the result of pandas' conversion of Categorical to a list, so not controlled by pyarrow. 

For `DictionaryArray.to_pandas` we return a pandas categorical dtype, which natively supports missing values (using -1 in the codes). So for this part, the `None` vs `NaN` is not relevant (outside of our control)

Reply:
I think that my question was mostly in terms of consistency.

Given
```
da = pa.DictionaryArray.from_arrays(pa.array([0, None]), pa.array(['a']))
```

If 
```
da.to_pandas()[1] == NaN
```

shouldn't
```
da.to_numpy()[1] == da.to_pandas()[1]
``` 

woundn't it be confusing that in one case you get back `NaN` and in the other `None`?

Reply:
I understand that it can be confusing, but the fact that `da.to_pandas()[1] == NaN` is the case, is a pandas behaviour / implementation detail (pandas doesn't use either NaN or None under the hood for a Categorical, but returns NaN when accessing the value. And, in a potential next version with pandas using nullable dtypes, this would no longer be true). 

I think the relevant cases where we would want consistency is for the different `to_numpy` conversions:

```
In [20]: pa.array(["a", "b", None]).to_numpy(zero_copy_only=False)
Out[20]: array(['a', 'b', None], dtype=object)

In [21]: pa.array(["a", "b", None]).dictionary_encode().to_numpy(zero_copy_only=False)
Out[21]: array(['a', 'b', 'b'], dtype=object)
```

where of course right now the last one is buggy (which is what this PR is solving), but so I would expect those two above to return the same in case of `to_numpy()`

Reply:
Ok, just wanted to make sure we didn't introduced inconsistencies in the APIs by chance.

I'll go in the direction of having the `to_numpy()` return the same values it would return with a normal `Array` and I'll check what's the impact of converting the `DictionaryArray`  to a plain `Array` before doing the conversion when there are nulls.